### ISSUE TITLE
Prover applies rule to undefined term

### ISSUE THREAD

**Issue by ehildenb** [kframework/kore#2798]
This file defines a new integer function called `#gas(_)` which represents positive infinity. Then we axiomatize it by saying that `#gas(_)` is not smaller than any number, nad is at least as big as every number.

```
requires "domains.md"

module TEST
    imports INT

    configuration <k> $PGM:Step </k>

    syntax Step ::= run ( Bool ) | done ( Bool )
 // --------------------------------------------
    rule <k> run(S) => done(S) ... </k>

    syntax Int ::= #gas ( Int ) [function, functional, no-evaluators]
 // -----------------------------------------------------------------
    rule #gas(_) <Int _  => false [simplification]
    rule _ <=Int #gas(_) => true  [simplification]
endmodule
```

This can be build with `kompile --backend haskell test.k --main-module TEST --syntax-module TEST`.

Then we add the claim which has a potentially undefined LHS (in the same file `test.k`, and run `kprove test.k --spec-module TEST-SPEC --def-module TEST`):

```
module TEST-SPEC
    imports TEST

    claim <k> run(_ /Int _ <=Int #gas(_)) => done(false) ... </k>
                   
endmodule
```

Here, I would expect the backend to _not_ apply the rule for `_ <=Int #gas(_)`, because it cannot show that the term `_ /Int _` is defined (not `#Bottom`). But instead it applies it and the proof goes through.

Versions:

```
% kompile --version
K version:    5.1.130
Build date:   Tue Aug 10 11:58:47 UTC 2021

% kore-exec --version
Kore version 0.51.0.0
Git:
  revision:     4dd38d94cbddccf36a6bb3821c5ae1f82c4ddfe7
  branch:       HEAD
  last commit:  UNKNOWN

% z3 --version
Z3 version 4.8.7 - 64 bit
```

**Comment by ana-pantilie** [comment]
We always assume that the LHS is defined, and generate definedness conditions for the LHS when simplifying it. I ran the spec and it ends up stuck here, as expected (since the RHS is `done(false)`): 
```
  #Not ( {
    _1
  #Equals
    0
  } )
#And
  <k>
    done ( true ) ~> _DotVar1 ~> .
  </k>
```
Notice the condition formed out of the the generated definedness condition.

**Comment by ehildenb** [comment]
Here is a modified definition (which introduces the potentially undefined term later in the execution than the LHS of the original claim):

```
requires "domains.md"

module TEST
    imports INT

    configuration <k> $PGM:Step </k>

    syntax Step ::= run ( Bool ) | done ( Bool ) | a ( Int , Int , Int )
 // --------------------------------------------------------------------
    rule <k> a(I1, I2, I3) => run(I1:Int /Int I2:Int <=Int #gas(I3:Int)) ... </k>
    rule <k> run(S) => done(S) ... </k>

    syntax Int ::= #gas ( Int ) [function, functional, no-evaluators]
 // -----------------------------------------------------------------
    rule #gas(_) <Int _  => false [simplification]
    rule _ <=Int #gas(_) => true  [simplification]
endmodule

module TEST-SPEC
    imports TEST

    claim <k> a(V1, V2, V3) => done(true)  ... </k>                         // fails
    claim <k> a(V1, V2, V3) => done(true)  ... </k> requires V2 =/=Int 0    // passes
    claim <k> a(V1, V2, V3) => done(false) ... </k>                         // fails
    claim <k> a(V1, V2, V3) => done(false) ... </k> requires V2 =/=Int 0    // fails
                                                                             
endmodule 
```

This one behaves how I expect, that is it will not prove the first claim because it cannot apply the lemma for `_ <=Int #gas(_)` because it does not know that the LHS of the `_<=Int_` is undefined.

However, the message it gives as output is rather confusing (for the first claim):

```
kore-exec: [388818] Error (ErrorDecidePredicateUnknown):
    Failed to decide predicate:
        /* Sfa */
        \and{_PREDICATE{}}(
            /* Sfa */
            \not{_PREDICATE{}}(
                /* Sfa */
                \equals{SortBool{}, _PREDICATE{}}(
                    /* Fl Fn D Sfa Cl */ \dv{SortBool{}}("true"),
                    /* Fn Sfa */
                    Lbl'Unds-LT-Eqls'Int'Unds'{}(
                        /* Fn Sfa */
                        Lbl'UndsSlsh'Int'Unds'{}(
                            /* Fl Fn D Sfa */ VarV1:SortInt{},
                            /* Fl Fn D Sfa */ VarV2:SortInt{}
                        ),
                        /* Fl Fn D Sfa */
                        Lbl'Hash'gas'LParUndsRParUnds'TEST'Unds'Int'Unds'Int{}(
                            /* Fl Fn D Sfa */ VarV3:SortInt{}
                        )
                    )
                )
            ),
            /* Sfa */                                                        
            \not{_PREDICATE{}}(                                               
                /* Spa */
                \equals{SortInt{}, _PREDICATE{}}(
                    /* Fl Fn D Sfa */ VarV2:SortInt{},
                    /* Fl Fn D Sfa Cl */ \dv{SortInt{}}("0")
                )
            )
        )
Created bug report: kore-exec.tar.gz
  #Not ( {
    V2
  #Equals
    0
  } )
#And
  <k>
    done ( V1 /Int V2 <=Int #gas ( V3 ) ) ~> _DotVar1 ~> .
  </k>
```

Here, it looks like it's outputting a final failure state where `V2 =/=Int 0`, which is the subset of states that it _should_ be able to show reaches the final configuration. I would expect it to output a final failure state with `V2 ==Int 0` instead, saying "for this case, I cannot show that the final state is reached". @ana-pantilie any idea what's going on here?

One thought I had is maybe the backend is assuming definedness of the term, which puts the extra restriction that `V2 =/=Int 0`, and what's actually failing is the check that the remainder is zero after adding this definedness condition?

**Comment by ehildenb** [comment]
Note that as long as the first claim continues to reliably _fail to be proved_, then I am happy, because it means we can write much simpler lemmas for `#gas(_)` (effectively cutting in half the number of lemmas we need https://github.com/kframework/evm-semantics/pull/1109). But I want to make sure we understand _why_ it fails to be proved, and all agree that it is working as intended there.

**Comment by ana-pantilie** [comment]
This one is failing because Z3 crashes on that predicate. I don't know why, but I'm pretty sure that it's not expected behavior.

**Comment by ehildenb** [comment]
Note that I think this relates to the design decision made a while ago which says "we expect users to write rules such that definedness is preserved between the LHS and RHS". I think though, this is too much to ask for users. For example, we have many lemmas that do not preserve this property (the above are examples). Another example would be `rule G -Int G => 0 [simplification]`, which is very intuitive for a user to write down, but if `G ==Int X /Int 0` the LHS is undefined and the RHS is defined.

**Comment by ehildenb** [comment]
I think at least, we should be checking for definedness of variables that show up on the LHS of the rule but not the RHS. If a variable shows up on both the LHS and RHS of a rule, it will preserved (un)definedness. But if it only shows up on the LHS, and not the RHS, then we could go from undefined to defined by applying the rule.

This check can be inserted by the frontend, so the backend doesn't need change. Does it sound correct to you @ana-pantilie ? Also, how do we phrase something like that to the bcakend? Do we just use `#Ceil`?

So, for example, if we had:

```
rule I -Int I => 0
rule _ <=Int #gas(_) => true
rule _ *Int 0 => 0
rule (_ *Int 0) +Int I => I
```

This would be translated to something like (is this correct?):

```
rule I -Int I => 0 requires #Ceil(I)
rule _0 <=Int #gas(_1) => true requires #Ceil(_0) andBool #Ceil(_1)
rule _0 *Int 0 => 0 requires #Ceil(_0)
rule (_0 *Int 0) +Int I => I requires #Ceil(_0)
```

Note that this is a design decision change, we used to say that "it's up to the user to preserve definedness", and now we will instead be saying "we'll check that definedness is preserved". I think the rules given here (which are used in basically any of the semantics which we do any proving with) are convincing arguments that we should allow the user write rules that do not preserve definedness, but only let them apply in the cases where definedness is preserved. It would be way too cumbersome for us to make sure that simple rules like this (and even more complicated rules in other cases) preserve definedness manually.

**Comment by ana-pantilie** [comment]
> I think this relates to the design decision made a while ago which says "we expect users to write rules such that definedness is preserved between the LHS and RHS".

I don't think this is related to that. Those simplification rules contained partial function symbols. A rule like `rule G -Int G => 0` doesn't contain any partial function symbols, and `G` can't ever be `#Bottom` because it's an element variable.

My understanding is that we shouldn't allow `G` to match anything which isn't `function-like` and known to be defined. I took a quick look at the code and this is exactly what we do right now. So, going back to your original example, we assume the configuration to be defined, and this gives us the definedness condition needed for that simplification rule to apply.

The reason why your second example doesn't pass is unknown to me. I would expect that after applying the semantic rule we would assume the configuration defined and be able to apply the simplification rule to the term. This might be a bug in equation application, and we've been seeing a few of them these days (our investigation is WIP).

To summarize, we always assume the configuration defined so we should be able to pass both those proofs. The reason we do this is because, in theory, we can branch on the condition that the configuration is defined. If it's not, then the proof is trivial. If it is, then we can continue rewriting. So basically we can only consider the branch where we assume it is. I believe it would actually be wrong to not make this assumption, because we could end up in many situations where the proof gets stuck because we have incomplete information about the configuration.
